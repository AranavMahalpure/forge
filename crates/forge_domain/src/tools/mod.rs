mod call;
mod choice;
mod definition;
mod dispatch;
mod name;
mod parser;
mod result;
mod tool;
mod usage;

pub use call::*;
pub use choice::*;
pub use definition::*;
pub use dispatch::*;
pub use name::*;
pub use parser::*;
pub use result::*;
pub use tool::*;
pub use usage::*;
