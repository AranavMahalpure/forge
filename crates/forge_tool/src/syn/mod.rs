mod validate;

pub use validate::{validate, Error};