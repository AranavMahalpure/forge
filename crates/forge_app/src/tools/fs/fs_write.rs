use std::path::Path;

use anyhow::Context;
use forge_display::{DiffFormat, VsCodeDiffConfig};
use forge_display::vscode_diff::create_streaming_diff_viewer;
use forge_domain::{ExecutableTool, NamedTool, ToolDescription, ToolName};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::Deserialize;

use crate::tools::syn;
use crate::tools::utils::assert_absolute_path;

#[derive(Deserialize, JsonSchema)]
pub struct FSWriteInput {
    /// The path of the file to write to (absolute path required)
    pub path: String,
    /// The content to write to the file. ALWAYS provide the COMPLETE intended
    /// content of the file, without any truncation or omissions. You MUST
    /// include ALL parts of the file, even if they haven't been modified.
    pub content: String,
}

/// File System Write Tool
///
/// Creates a new file at the specified path with the provided content.
/// 
/// # Description
/// 
/// This tool allows creating or overwriting files at absolute paths. It:
/// - Requires absolute paths for file locations
/// - Creates any missing intermediary directories automatically
/// - Validates syntax for supported file types
/// - Shows diff between old and new content in terminal and VS Code
/// 
/// # Important
/// 
/// DO NOT attempt to use this tool to move or rename files, use the shell tool instead.
#[derive(ToolDescription)]
pub struct FSWrite;

impl NamedTool for FSWrite {
    /// Returns the name of the tool as a ToolName
    fn tool_name() -> ToolName {
        ToolName::new("tool_forge_fs_create")
    }
}

#[async_trait::async_trait]
impl ExecutableTool for FSWrite {
    type Input = FSWriteInput;

    /// Creates a file at the specified path with the provided content.
    /// 
    /// # Parameters
    /// 
    /// * `input` - Contains the path and content for the file to be created
    /// 
    /// # Returns
    /// 
    /// * `Ok(String)` - Success message with bytes written and path
    /// * `Err(anyhow::Error)` - If file creation fails for any reason
    /// 
    /// # Notes
    /// 
    /// This implementation:
    /// - Creates parent directories automatically if they don't exist
    /// - Validates syntax for supported languages (Rust, JavaScript, TypeScript, etc.)
    /// - Shows both terminal and VS Code diff visualization of changes
    /// - Supports animated diffs for better change visualization
    /// - Displays appropriate warnings for syntax issues
    /// - Requires absolute paths for security and clarity
    async fn call(&self, input: Self::Input) -> anyhow::Result<String> {
        // Validate absolute path requirement
        let path = Path::new(&input.path);
        assert_absolute_path(path)?;

        // Validate file content if it's a supported language file
        let syntax_warning = syn::validate(&input.path, &input.content);

        // Create parent directories if they don't exist
        if let Some(parent) = Path::new(&input.path).parent() {
            tokio::fs::create_dir_all(parent)
                .await
                .with_context(|| format!("Failed to create directories: {}", input.path))?;
        }

        // record the file content before they're modified
        let old_content = if path.is_file() {
            // if file already exists, we should be able to read it.
            tokio::fs::read_to_string(path).await?
        } else {
            // if file doesn't exist, we should record it as an empty string.
            "".to_string()
        };

        // Write file only after validation passes and directories are created
        tokio::fs::write(&input.path, &input.content).await?;

        let mut result = format!(
            "Successfully wrote {} bytes to {}",
            input.content.len(),
            input.path
        );
        if let Some(warning) = syntax_warning {
            result.push_str("\nWarning: ");
            result.push_str(&warning.to_string());
        }

        // Initialize VS Code diff viewer for streaming updates with animation
        let vscode_config = VsCodeDiffConfig {
            animated_diff: true,          // Enable animated diff
            animation_delay_ms: 80,       // Slightly faster animation for better user experience
            update_frequency_ms: 50,      // More responsive updates
            min_change_percent: 0.1,      // More sensitive to changes
            ..VsCodeDiffConfig::default()
        };
        let mut vs_diff_viewer = match create_streaming_diff_viewer(path, &old_content, vscode_config).await {
            Ok(viewer) => Some(viewer),
            Err(e) => {
                eprintln!("Warning: Failed to create streaming VS Code diff viewer: {}", e);
                None
            }
        };
        
        // Read the modified file content
        let new_content = tokio::fs::read_to_string(path).await?;
        
        // Stream diff line by line to the terminal
        DiffFormat::stream(path.to_path_buf(), &old_content, &new_content, |line| {
            // Update VS Code diff viewer for each line of terminal output
            if let Some(viewer) = &mut vs_diff_viewer {
                if let Err(e) = tokio::task::block_in_place(|| {
                    tokio::runtime::Handle::current().block_on(async {
                        viewer.update(&new_content).await
                    })
                }) {
                    eprintln!("Warning: Failed to update VS Code diff: {}", e);
                }
            }
            
            // Print the line to terminal
            println!("{}", line);
        });
        
        // Finalize the VS Code diff view
        if let Some(mut viewer) = vs_diff_viewer {
            if let Err(e) = viewer.finalize(&new_content).await {
                eprintln!("Warning: Failed to finalize VS Code diff: {}", e);
            }
        }

        Ok(result)
    }
}

#[cfg(test)]
mod test {
    use std::path::Path;

    use pretty_assertions::assert_eq;
    use tokio::fs;

    use super::*;
    use crate::tools::utils::TempDir;

    /// Asserts that a path exists in the filesystem.
    /// 
    /// # Parameters
    /// 
    /// * `path` - The path to check for existence
    /// 
    /// # Panics
    /// 
    /// If the path does not exist
    async fn assert_path_exists(path: impl AsRef<Path>) {
        assert!(fs::metadata(path).await.is_ok(), "Path should exist");
    }

    #[tokio::test]
    async fn test_fs_write_success() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.txt");
        let content = "Hello, World!";

        let fs_write = FSWrite;
        let output = fs_write
            .call(FSWriteInput {
                path: file_path.to_string_lossy().to_string(),
                content: content.to_string(),
            })
            .await
            .unwrap();

        assert!(output.contains("Successfully wrote"));
        assert!(output.contains(&file_path.display().to_string()));
        assert!(output.contains(&content.len().to_string()));

        // Verify file was actually written
        let content = fs::read_to_string(&file_path).await.unwrap();
        assert_eq!(content, "Hello, World!")
    }

    #[tokio::test]
    async fn test_fs_write_invalid_rust() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.rs");

        let fs_write = FSWrite;
        let result = fs_write
            .call(FSWriteInput {
                path: file_path.to_string_lossy().to_string(),
                content: "fn main() { let x = ".to_string(),
            })
            .await;

        let output = result.unwrap();
        assert!(output.contains("Warning:"));
    }

    #[tokio::test]
    async fn test_fs_write_valid_rust() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.rs");

        let fs_write = FSWrite;
        let content = "fn main() { let x = 42; }";
        let result = fs_write
            .call(FSWriteInput {
                path: file_path.to_string_lossy().to_string(),
                content: content.to_string(),
            })
            .await;

        let output = result.unwrap();
        assert!(output.contains("Successfully wrote"));
        assert!(output.contains(&file_path.display().to_string()));
        assert!(output.contains(&content.len().to_string()));
        assert!(!output.contains("Warning:"));
        // Verify file contains valid Rust code
        let content = fs::read_to_string(&file_path).await.unwrap();
        assert_eq!(content, "fn main() { let x = 42; }");
    }

    #[tokio::test]
    async fn test_fs_write_single_directory_creation() {
        let temp_dir = TempDir::new().unwrap();
        let nested_path = temp_dir.path().join("new_dir").join("test.txt");
        let content = "Hello from nested file!";

        let fs_write = FSWrite;
        let result = fs_write
            .call(FSWriteInput {
                path: nested_path.to_string_lossy().to_string(),
                content: content.to_string(),
            })
            .await
            .unwrap();

        assert!(result.contains("Successfully wrote"));
        // Verify both directory and file were created
        assert_path_exists(&nested_path).await;
        assert_path_exists(nested_path.parent().unwrap()).await;

        // Verify content
        let written_content = fs::read_to_string(&nested_path).await.unwrap();
        assert_eq!(written_content, content);
    }

    #[tokio::test]
    async fn test_fs_write_deep_directory_creation() {
        let temp_dir = TempDir::new().unwrap();
        let deep_path = temp_dir
            .path()
            .join("level1")
            .join("level2")
            .join("level3")
            .join("deep.txt");
        let content = "Deep in the directory structure";

        let fs_write = FSWrite;
        let result = fs_write
            .call(FSWriteInput {
                path: deep_path.to_string_lossy().to_string(),
                content: content.to_string(),
            })
            .await
            .unwrap();

        assert!(result.contains("Successfully wrote"));

        // Verify entire path was created
        assert_path_exists(&deep_path).await;
        let mut current = deep_path.parent().unwrap();
        while current != temp_dir.path() {
            assert_path_exists(current).await;
            current = current.parent().unwrap();
        }

        // Verify content
        let written_content = fs::read_to_string(&deep_path).await.unwrap();
        assert_eq!(written_content, content);
    }

    #[tokio::test]
    async fn test_fs_write_with_different_separators() {
        let temp_dir = TempDir::new().unwrap();

        // Use forward slashes regardless of platform
        let path_str = format!("{}/dir_a/dir_b/file.txt", temp_dir.path().to_string_lossy());
        let content = "Testing path separators";

        let fs_write = FSWrite;
        let result = fs_write
            .call(FSWriteInput { path: path_str, content: content.to_string() })
            .await
            .unwrap();

        assert!(result.contains("Successfully wrote"));

        // Convert to platform path and verify
        let platform_path = Path::new(&temp_dir.path())
            .join("dir_a")
            .join("dir_b")
            .join("file.txt");

        assert_path_exists(&platform_path).await;

        // Verify content
        let written_content = fs::read_to_string(&platform_path).await.unwrap();
        assert_eq!(written_content, content);
    }

    #[tokio::test]
    async fn test_fs_write_relative_path() {
        let fs_write = FSWrite;
        let result = fs_write
            .call(FSWriteInput {
                path: "relative/path/file.txt".to_string(),
                content: "test content".to_string(),
            })
            .await;

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Path must be absolute"));
    }
}
