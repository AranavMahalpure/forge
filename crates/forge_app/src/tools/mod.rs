mod fetch;
mod fs;
mod patch;
mod shell;
mod syn;
mod think;
mod utils;

use std::sync::Arc;

use fetch::Fetch;
use forge_domain::Tool;
use fs::*;
use patch::*;
use shell::Shell;
use think::Think;

use crate::{EnvironmentService, Infrastructure};

pub fn tools<F: Infrastructure>(infra: Arc<F>) -> Vec<Tool> {
    let env = infra.environment_service().get_environment();
    vec![
        FSRead.into(),
        FSWrite::new(infra.clone()).into(),
        FSRemove.into(),
        FSList::default().into(),
        FSSearch.into(),
        FSFileInfo.into(),
        // TODO: once ApplyPatchJson is stable we can delete ApplyPatch
        // ApplyPatch::new(infra.clone()).into(),
        ApplyPatchJson::new(infra).into(),
        Shell::new(env.clone()).into(),
        Think::default().into(),
        Fetch::default().into(),
    ]
}

#[cfg(test)]
mod tests {
    use std::path::{Path, PathBuf};

    use bytes::Bytes;
    use forge_domain::{Environment, Point, Query, Suggestion};
    use forge_snaps::{FileSnapshotService, SnapshotInfo, SnapshotMetadata};

    use super::*;
    use crate::{EmbeddingService, FileExist, FileReadService, FileWriteService, VectorIndex};

    /// Create a default test environment
    fn stub() -> Stub {
        Stub {
            env: Environment {
                os: std::env::consts::OS.to_string(),
                cwd: std::env::current_dir().unwrap_or_default(),
                home: Some("/".into()),
                shell: if cfg!(windows) {
                    "cmd.exe".to_string()
                } else {
                    "/bin/sh".to_string()
                },
                base_path: PathBuf::new(),
                qdrant_key: Default::default(),
                qdrant_cluster: Default::default(),
                pid: std::process::id(),
                provider_url: Default::default(),
                provider_key: Default::default(),
                openai_key: Default::default(),
            },
        }
    }

    struct Stub {
        env: Environment,
    }

    #[async_trait::async_trait]
    impl EmbeddingService for Stub {
        async fn embed(&self, _text: &str) -> anyhow::Result<Vec<f32>> {
            unimplemented!()
        }
    }

    #[async_trait::async_trait]
    impl EnvironmentService for Stub {
        fn get_environment(&self) -> Environment {
            self.env.clone()
        }
    }
    #[async_trait::async_trait]
    impl FileReadService for Stub {
        async fn read(&self, _path: &Path) -> anyhow::Result<Bytes> {
            unimplemented!()
        }
    }

    #[async_trait::async_trait]
    impl FileWriteService for Stub {
        async fn write(&self, _: &Path, _: Bytes) -> anyhow::Result<()> {
            unimplemented!()
        }
    }
    #[async_trait::async_trait]
    impl VectorIndex<Suggestion> for Stub {
        async fn store(&self, _information: Point<Suggestion>) -> anyhow::Result<()> {
            unimplemented!()
        }

        async fn search(&self, _query: Query) -> anyhow::Result<Vec<Point<Suggestion>>> {
            unimplemented!()
        }
    }

    #[async_trait::async_trait]
    impl FileSnapshotService for Stub {
        fn snapshot_dir(&self) -> PathBuf {
            todo!()
        }

        async fn create_snapshot(&self, _: &Path) -> anyhow::Result<SnapshotInfo> {
            todo!()
        }

        async fn list_snapshots(&self, _: &Path) -> anyhow::Result<Vec<SnapshotInfo>> {
            todo!()
        }

        async fn restore_by_timestamp(&self, _: &Path, _: &str) -> anyhow::Result<()> {
            todo!()
        }

        async fn restore_by_index(&self, _: &Path, _: isize) -> anyhow::Result<()> {
            todo!()
        }

        async fn restore_previous(&self, _: &Path) -> anyhow::Result<()> {
            todo!()
        }

        async fn get_snapshot_by_timestamp(&self, _: &Path, _: &str) -> anyhow::Result<SnapshotMetadata> {
            todo!()
        }

        async fn get_snapshot_by_index(
            &self,
            _: &Path,
            _: isize,
        ) -> anyhow::Result<SnapshotMetadata> {
            todo!()
        }

        async fn purge_older_than(&self, _: u32) -> anyhow::Result<usize> {
            todo!()
        }
    }

    #[async_trait::async_trait]
    impl FileExist for Stub {
        async fn exist(&self, _: &Path) -> anyhow::Result<bool> {
            todo!()
        }
    }

    #[async_trait::async_trait]
    impl Infrastructure for Stub {
        type EnvironmentService = Stub;
        type FileReadService = Stub;
        type FileWriteService = Stub;
        type VectorIndex = Stub;
        type EmbeddingService = Stub;
        type FileSnapshotService = Stub;
        type FileExist = Stub;

        fn environment_service(&self) -> &Self::EnvironmentService {
            self
        }

        fn file_read_service(&self) -> &Self::FileReadService {
            self
        }

        fn file_write_service(&self) -> &Self::FileWriteService {
            self
        }

        fn vector_index(&self) -> &Self::VectorIndex {
            self
        }

        fn embedding_service(&self) -> &Self::EmbeddingService {
            self
        }

        fn file_snapshot_service(&self) -> &Self::FileSnapshotService {
            self
        }

        fn file_exist_service(&self) -> &Self::FileExist {
            todo!()
        }
    }

    #[test]
    fn test_tool_description_length() {
        const MAX_DESCRIPTION_LENGTH: usize = 1024;

        println!("\nTool description lengths:");

        let mut any_exceeded = false;
        let stub = Arc::new(stub());
        for tool in tools(stub.clone()) {
            let desc_len = tool.definition.description.len();
            println!(
                "{:?}: {} chars {}",
                tool.definition.name,
                desc_len,
                if desc_len > MAX_DESCRIPTION_LENGTH {
                    "(!)"
                } else {
                    ""
                }
            );

            if desc_len > MAX_DESCRIPTION_LENGTH {
                any_exceeded = true;
            }
        }

        assert!(
            !any_exceeded,
            "One or more tools exceed the maximum description length of {}",
            MAX_DESCRIPTION_LENGTH
        );
    }
}
