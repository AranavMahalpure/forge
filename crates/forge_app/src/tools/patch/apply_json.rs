use std::path::Path;

// No longer using dissimilar for fuzzy matching
use forge_domain::{ExecutableTool, NamedTool, ToolDescription, ToolName};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use thiserror::Error;
use tokio::fs;

use crate::tools::syn;
use crate::tools::utils::assert_absolute_path;

// Removed fuzzy matching threshold as we only use exact matching now

/// A match found in the source text. Represents a range in the source text that
/// can be used for extraction or replacement operations. Stores the position
/// and length to allow efficient substring operations.
#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
struct Range {
    /// Starting position of the match in source text
    start: usize,
    /// Length of the matched text
    length: usize,
}

impl Range {
    /// Create a new match from a start position and length
    fn new(start: usize, length: usize) -> Self {
        Self { start, length }
    }

    /// Get the end position (exclusive) of this match
    fn end(&self) -> usize {
        self.start + self.length
    }

    /// Try to find an exact match in the source text
    fn find_exact(source: &str, search: &str) -> Option<Self> {
        source
            .find(search)
            .map(|start| Self::new(start, search.len()))
    }

    // Fuzzy matching removed - we only use exact matching
}

impl From<Range> for std::ops::Range<usize> {
    fn from(m: Range) -> Self {
        m.start..m.end()
    }
}

// MatchSequence struct and implementation removed - we only use exact matching

#[derive(Debug, Error)]
enum Error {
    #[error("Failed to read/write file: {0}")]
    FileOperation(#[from] std::io::Error),
    #[error("Could not find match for search text: {0}")]
    NoMatch(String),
    #[error("Could not find swap target text: {0}")]
    NoSwapTarget(String),
}

fn apply_replacement(source: String, replacement: &Replacement) -> Result<String, Error> {
    let search = replacement.search.as_str();

    // Handle empty search string - only certain operations make sense here
    if replacement.search.is_empty() {
        return match &replacement.operation {
            // Append to the end of the file
            Operation::Append(content) => Ok(format!("{}{}", source, content)),
            // Prepend to the beginning of the file
            Operation::Prepend(content) => Ok(format!("{}{}", content, source)),
            // Replace is equivalent to completely replacing the file
            Operation::Replace(content) => Ok(content.clone()),
            // Swap doesn't make sense with empty search - keep source unchanged
            Operation::Swap(_) => Ok(source),
        };
    }

    // Find the exact match to operate on
    let patch = Range::find_exact(&source, search)
        .ok_or_else(|| Error::NoMatch(replacement.search.clone()))?;

    // Apply the operation based on its type
    match &replacement.operation {
        // Prepend content before the matched text
        Operation::Prepend(content) => Ok(format!(
            "{}{}{}",
            &source[..patch.start],
            content,
            &source[patch.start..]
        )),

        // Append content after the matched text
        Operation::Append(content) => Ok(format!(
            "{}{}{}",
            &source[..patch.end()],
            content,
            &source[patch.end()..]
        )),

        // Replace matched text with new content
        Operation::Replace(content) => Ok(format!(
            "{}{}{}",
            &source[..patch.start],
            content,
            &source[patch.end()..]
        )),

        // Swap with another text in the source
        Operation::Swap(target) => {
            // Find the target text to swap with
            let target_patch = Range::find_exact(&source, target)
                .ok_or_else(|| Error::NoSwapTarget(target.clone()))?;

            // Handle the case where patches overlap
            if (patch.start <= target_patch.start && patch.end() > target_patch.start)
                || (target_patch.start <= patch.start && target_patch.end() > patch.start)
            {
                // For overlapping ranges, we just do an ordinary replacement
                return Ok(format!(
                    "{}{}{}",
                    &source[..patch.start],
                    target,
                    &source[patch.end()..]
                ));
            }

            // We need to handle different ordering of patches
            if patch.start < target_patch.start {
                // Original text comes first
                Ok(format!(
                    "{}{}{}{}{}",
                    &source[..patch.start],
                    target,
                    &source[patch.end()..target_patch.start],
                    &source[patch.start..patch.end()],
                    &source[target_patch.end()..]
                ))
            } else {
                // Target text comes first
                Ok(format!(
                    "{}{}{}{}{}",
                    &source[..target_patch.start],
                    &source[patch.start..patch.end()],
                    &source[target_patch.end()..patch.start],
                    target,
                    &source[patch.end()..]
                ))
            }
        }
    }
}

/// Operation types that can be performed on matched text
#[derive(Deserialize, Serialize, JsonSchema, Debug, Clone, PartialEq)]
#[serde(tag = "type", content = "content")]
pub enum Operation {
    /// Prepend content before the matched text
    Prepend(String),

    /// Append content after the matched text
    Append(String),

    /// Replace the matched text with new content
    Replace(String),

    /// Swap the matched text with another text (search for the second text and
    /// swap them)
    Swap(String),
}

/// A single search and operation pair
#[derive(Deserialize, Serialize, JsonSchema, Debug, Clone)]
pub struct Replacement {
    /// The text to search for in the source. If empty, operation applies to the
    /// end of the file.
    pub search: String,

    /// The operation to perform on the matched text
    pub operation: Operation,
}

#[derive(Deserialize, JsonSchema)]
pub struct ApplyPatchJsonInput {
    pub path: String,
    pub replacement: Replacement,
}

/// Performs a single text operation (prepend, append, replace, swap, delete) on
/// matched text in a file. The operation is applied to the first match found in
/// the text.
#[derive(ToolDescription)]
pub struct ApplyPatchJson;

impl NamedTool for ApplyPatchJson {
    fn tool_name() -> ToolName {
        ToolName::new("tool_forge_patch_v2")
    }
}

/// Format the modified content as XML with optional syntax warning
fn format_output(path: &str, content: &str, warning: Option<&str>) -> String {
    if let Some(w) = warning {
        format!(
            "<file_content\n  path=\"{}\"\n  syntax_checker_warning=\"{}\">\n{}</file_content>\n",
            path, w, content
        )
    } else {
        format!(
            "<file_content path=\"{}\">\n{}\n</file_content>\n",
            path,
            content.trim_end()
        )
    }
}

/// Process the file modification and return the formatted output
async fn process_file_modifications(
    path: &Path,
    replacement: &Replacement,
) -> Result<String, Error> {
    let content = fs::read_to_string(path).await?;
    let content = apply_replacement(content, replacement)?;
    fs::write(path, &content).await?;

    let warning = syn::validate(path, &content).map(|e| e.to_string());
    Ok(format_output(
        path.to_string_lossy().as_ref(),
        &content,
        warning.as_deref(),
    ))
}

#[async_trait::async_trait]
impl ExecutableTool for ApplyPatchJson {
    type Input = ApplyPatchJsonInput;

    async fn call(&self, input: Self::Input) -> anyhow::Result<String> {
        let path = Path::new(&input.path);
        assert_absolute_path(path)?;

        Ok(process_file_modifications(path, &input.replacement).await?)
    }
}

#[cfg(test)]
mod test {
    use std::fmt::{self, Display};

    use super::*;

    // Test helpers
    #[derive(Debug)]
    struct PatchTest {
        initial: String,
        replacement: Option<Replacement>,
        result: Option<String>,
    }

    impl fmt::Display for PatchTest {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            let replacement = match &self.replacement {
                Some(r) => r.to_string(),
                None => String::new(),
            };
            // Use the original tag "REPLACEMENTS" for backward compatibility
            write!(
                f,
                "\n<!-- INITIAL -->{}\n<!-- REPLACEMENTS -->{}\n<!-- FINAL -->{}",
                self.initial,
                replacement,
                self.result
                    .as_ref()
                    .expect("Test must be executed before display")
            )
        }
    }

    impl PatchTest {
        fn new(initial: impl ToString) -> Self {
            PatchTest {
                initial: initial.to_string(),
                replacement: None,
                result: Default::default(),
            }
        }

        /// Replace matched text with new content
        fn replace(mut self, search: impl ToString, content: impl ToString) -> Self {
            self.replacement = Some(Replacement {
                search: search.to_string(),
                operation: Operation::Replace(content.to_string()),
            });
            self
        }

        /// Prepend content before matched text
        fn prepend(mut self, search: impl ToString, content: impl ToString) -> Self {
            self.replacement = Some(Replacement {
                search: search.to_string(),
                operation: Operation::Prepend(content.to_string()),
            });
            self
        }

        /// Append content after matched text
        fn append(mut self, search: impl ToString, content: impl ToString) -> Self {
            self.replacement = Some(Replacement {
                search: search.to_string(),
                operation: Operation::Append(content.to_string()),
            });
            self
        }

        /// Swap matched text with target text
        fn swap(mut self, search: impl ToString, target: impl ToString) -> Self {
            self.replacement = Some(Replacement {
                search: search.to_string(),
                operation: Operation::Swap(target.to_string()),
            });
            self
        }

        // TODO: tests don't need to write files to disk
        fn execute(mut self) -> Result<Self, Error> {
            if let Some(replacement) = &self.replacement {
                self.result = Some(apply_replacement(self.initial.clone(), replacement)?);
                Ok(self)
            } else {
                // No replacement specified - return original content
                self.result = Some(self.initial.clone());
                Ok(self)
            }
        }
    }

    // For backward compatibility with existing snapshots, format all operations
    // using the original SEARCH/REPLACE format where possible
    impl Display for Replacement {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            match &self.operation {
                // For backward compatibility with tests, use REPLACE instead of specialized tags
                Operation::Prepend(content) => write!(
                    f,
                    "\n<!-- SEARCH -->{}\n<!-- PREPEND -->{}",
                    self.search, content
                ),
                Operation::Append(content) => write!(
                    f,
                    "\n<!-- SEARCH -->{}\n<!-- APPEND -->{}",
                    self.search, content
                ),
                Operation::Replace(content) => write!(
                    f,
                    "\n<!-- SEARCH -->{}\n<!-- REPLACE -->{}",
                    self.search, content
                ),
                // For new operations that don't have existing snapshots, use new tags
                Operation::Swap(target) => write!(
                    f,
                    "\n<!-- SEARCH -->{}\n<!-- SWAP -->{}",
                    self.search, target
                ),
            }
        }
    }

    /*
     * Basic Operations
     * Tests basic functionality like exact matches, empty inputs, and simple
     * cases
     */
    #[test]
    fn exact_match_single_word() {
        let actual = PatchTest::new("Hello World")
            .replace("World", "Forge")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn append_empty_search() {
        let actual = PatchTest::new("foo").replace("", " bar").execute().unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn delete_empty_replace() {
        let actual = PatchTest::new("foo bar baz")
            .replace("bar ", "")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Single Replacement Behavior
     * Tests behavior when single replacements are performed
     */
    #[test]
    fn replaces_only_first_match() {
        let actual = PatchTest::new("foo bar foo")
            .replace("foo", "baz")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Exact Matching Behavior
     * Tests the exact matching behavior
     */

    #[test]
    fn no_match_with_hyphenated_pattern() {
        let result = PatchTest::new("abc foobar pqr")
            .replace("foo-bar", "replaced") // Should NOT match "foobar" with hyphen
            .execute();
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Could not find match"));
    }

    #[test]
    fn no_match_with_prefix() {
        let result = PatchTest::new("foox baz foo")
            .replace("afoo", "bar") // Should NOT match "foox" or "foo" with "a" prefix
            .execute();
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Could not find match"));
    }

    #[test]
    fn matches_only_first_occurrence() {
        let actual = PatchTest::new("foo foox foo")
            .replace("foo", "bar") // Should replace first exact "foo" only
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Unicode and Special Characters
     * Tests handling of non-ASCII text and special characters
     */
    #[test]
    fn unicode_characters() {
        let actual = PatchTest::new("Hello 世界")
            .replace("世界", "🌍")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn emoji_replacement() {
        let actual = PatchTest::new("Hi 👋, how are you?")
            .replace("👋", "👍")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn mixed_unicode_ascii() {
        let actual = PatchTest::new("Hello 世界 World")
            .replace("世界 World", "地球")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Whitespace Handling
     * Tests preservation of whitespace, indentation, and line endings
     */
    #[test]
    fn preserve_indentation() {
        let actual = PatchTest::new("    indented\n        more indented")
            .replace("indented", "text")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn tab_characters() {
        let actual = PatchTest::new("no_tab\thas_tab")
            .replace("has_tab", "replaced")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn multiple_line_endings() {
        let actual = PatchTest::new("line1\nline2\rline3\r\nline4")
            .replace("line2", "replaced")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Error Cases
     * Tests error handling and edge cases
     */
    #[test]
    fn single_replacement_only() {
        let actual = PatchTest::new("outer inner outer")
            .replace("outer inner outer", "modified inner")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Complex Replacements
     * Tests complicated scenarios like nested and overlapping matches
     */

    /*
     * Operation Type Tests
     * Tests for each specific operation type
     */

    #[test]
    fn prepend_operation() {
        let actual = PatchTest::new("Hello World!")
            .prepend("Hello", "Greetings, ")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn append_operation() {
        let actual = PatchTest::new("Hello World")
            .append("World", "!")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn replace_operation() {
        let actual = PatchTest::new("Hello World")
            .replace("World", "Universe")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn swap_operation() {
        let actual = PatchTest::new("Hello World")
            .swap("Hello", "World")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn swap_with_overlap() {
        let actual = PatchTest::new("Hello World")
            .swap("Hello W", "orld")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn swap_target_not_found() {
        let result = PatchTest::new("Hello World")
            .swap("Hello", "Universe")
            .execute();
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Could not find swap target"));
    }

    #[test]
    fn empty_search_prepend() {
        let actual = PatchTest::new("Hello World")
            .prepend("", "Start: ")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn empty_search_append() {
        let actual = PatchTest::new("Hello World")
            .append("", " End")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn empty_search_replace() {
        let actual = PatchTest::new("Hello World")
            .replace("", "Completely New Content")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Error Cases
     * Tests error handling and validation
     */

    #[test]
    fn delete_single_line_only() {
        let actual = PatchTest::new("line1\nline2\nline1\nline3")
            .replace("line1\n", "") // Using replace with empty string instead of delete for compatibility
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn empty_search_text() {
        let actual = PatchTest::new("").replace("", "foo").execute().unwrap(); // Using replace instead of append for compatibility
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn no_match_found() {
        let result = PatchTest::new("foo").replace("bar", "baz").execute();
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Could not find match"));
    }

    #[test]
    fn non_matching_pattern() {
        let result = PatchTest::new("fo").replace("foo", "bar").execute();
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Could not find match"));
    }
}
