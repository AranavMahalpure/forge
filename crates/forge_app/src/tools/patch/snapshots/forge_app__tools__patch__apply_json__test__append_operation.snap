---
source: crates/forge_app/src/tools/patch/apply_json.rs
expression: actual
---
<!-- INITIAL -->Hello World
<!-- REPLACEMENTS -->
<!-- SEARCH -->World
<!-- APPEND -->!
<!-- FINAL -->Hello World!
